woocommerce/woocommerce
# Use first meta value instead of last to be consistent with WP_Post.

### Submission Review Guidelines:

- I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
- I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/). 
- Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:



In several cases, we will have multiple meta values for supposedly unique meta keys, such as order_key, date_paid, etc, since in the post meta table, the unique constraint on the post_id <> meta_key combination cannot be enforced. When migrated to HPOS, unique constrained is enforced in the sense that these keys are not mapped to flat columns, which can only hold one value per row.

Currently, when multi-record are encountered, we select the last one, however, this is inconsistent with how WP core treats the case of multiple meta values for supposedly unique meta keys, which is to select the first value. This PR makes the HPOS migration consistent with the WP core's behavior in selecting the first meta value instead of the last one when there are duplicate meta.

Closes #37660 



### How to test the changes in this Pull Request:




Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Set the posts to authoritative and sync to off.
2. Create a new order, note it's ID. Make sure it has some value as the billing first name.
3. Add another meta entry for a supposedly unique meta key, such as `_billing_first_name`. You can do it from CLI like so: `wp post meta add XXXX _billing_first_name testname` where XXXX is the ID of the order created in the previous step.
4. Migrate the order to HPOS.
5. Open the order again with HPOS authoritative, you should still see the original first value instead of the `testname`.
